java/build/bazel/tests/integration/WorkspaceDriver.java

```diff
@@ -64,6 +63,10 @@ public static void setUpClass() throws IOException {
     javaToolchain = javaToolchainFromProperties();
   }
 
+  public static Properties globalBazelProperties() {
+    return Optional.ofNullable(properties)
+        .orElseThrow(() -> new IllegalStateException("Need to call setupClass() before calling globalBazelProperties()"));
```

Comment:
Can you just call setupClass for them instead of throwing and make setupClass be idempotent? You could use a private static "alreadyInitialized" to achieve that.

Reply:
I can but I don’t think it’s good design.
We’re not talking about runtime of web micro services. Worst case their test will fail, they will fix it and continue.
Why do you think implicit is better?

Reply:
`setUpClass` takes no arguments and therefore has no configuration. You're throwing an exception that tells the user to call a method. Why put that on the user when you have already detected that the method wasn't called? Your API can have one function call instead of two, why not opt for one?

Reply:
Because it’s very implicit that doing a “get” will trigger a big, IO including, setup.
https://stackoverflow.com/questions/28382151/command-query-separation-commands-must-return-void

Also remember that my intuition is that no actual user will need this getter. Maybe we should make it package protected.

Reply:
If there is only a small chance it'll be needed let's make it protected with a comment

Reply:
done